SampleViewer/DownloadsManager.cpp

```diff
@@ -0,0 +1,933 @@
+// [Legal]
+// Copyright 2025 Esri.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// [Legal]
+
+#include "pch.hpp"
+
+// ArcGIS Maps SDK headers
+#include "ErrorException.h"
+#include "MapTypes.h"
+#include "NetworkRequestProgress.h"
+#include "Portal.h"
+#include "PortalItem.h"
+#include "PortalTypes.h"
+
+// Qt headers
+#include <QDateTime>
+#include <QDebug>
+#include <QDir>
+#include <QDirIterator>
+#include <QFile>
+#include <QFileInfo>
+#include <QJsonArray>
+#include <QJsonDocument>
+#include <QJsonObject>
+#include <QSet>
+#include <QStandardPaths>
+#include <QTimer>
+#include <QUrl>
+
+// STL headers
+#include <algorithm>
+
+// Other headers
+#include "DataItem.h"
+#include "DataItemListModel.h"
+#include "DownloadsManager.h"
+#include "OfflineDataProjectsModel.h"
+#include "Sample.h"
+#include "SampleListModel.h"
+#include "ZipHelper.h"
+
+using namespace Esri::ArcGISRuntime;
+
+DownloadsManager::DownloadsManager(QObject* parent) :
+  QObject(parent),
+  m_offlineDataProjects(new OfflineDataProjectsModel(this))
+{
+  m_progressUpdateTimer = new QTimer(this);
+  m_progressUpdateTimer->setInterval(PROGRESS_UPDATE_INTERVAL_MS);
+  connect(m_progressUpdateTimer, &QTimer::timeout, this, [this]()
+  {
+    // Calculate overall progress across all downloading items
+    if (m_dataItemProgress.isEmpty())
+    {
+      setDownloadProgress(0.0);
+      return;
+    }
+
+    double totalProgress = 0.0;
+    for (const double& progress : m_dataItemProgress)
+    {
+      totalProgress += progress;
+    }
+    setDownloadProgress(totalProgress / m_dataItemProgress.size());
+
+    updateOfflineDataProjects();
+  });
+}
+
+DownloadsManager::~DownloadsManager()
+{
+}
+
+void DownloadsManager::initialize(Portal* portal, const QString& homePath)
+{
+  if (!portal)
+  {
+    qWarning() << "DownloadsManager::initialize: Portal is null!";
+    return;
+  }
+
+  if (homePath.isEmpty())
+  {
+    qWarning() << "DownloadsManager::initialize: Home path is empty!";
+    return;
+  }
+
+  m_portal = portal;
+  m_homePath = homePath;
+}
+
+int DownloadsManager::countDownloadedItems(Sample* sample) const
+{
+  if (!sample || sample->dataItems()->size() == 0)
+    return 0;
+
+  int downloadedItems = 0;
+  const int totalItems = sample->dataItems()->size();
+  for (int j = 0; j < totalItems; ++j)
```

Comment:
optional, but typically j and k are used for inner, nested for loops. As soon as I saw 'j', I just assumed this was a nested loop. I would consider ol' trusty 'i' here and below.